File: course_4/module_1/examples/floyd_warshall.py
class FloydWarshall:
    """
    A class to represent and solve the All-Pairs Shortest Paths problem using the Floyd-Warshall algorithm.
    """

    def __init__(self, vertices):
        """
        Initialize the FloydWarshall object.

        :param vertices: The number of vertices in the graph.
        """
        self.vertices = vertices
        # Initialize the distance matrix with infinity (no direct path).
        self.distances = [[float('inf')] * vertices for _ in range(vertices)]
        # Initialize the next matrix to reconstruct paths.
        self.next = [[None] * vertices for _ in range(vertices)]
        # Distance to self is zero.
        for i in range(vertices):
            self.distances[i][i] = 0

    def add_edge(self, u, v, weight):
        """
        Add an edge to the graph with a specified weight.

        :param u: The source vertex (0-indexed).
        :param v: The destination vertex (0-indexed).
        :param weight: The weight of the edge.
        :raises ValueError: If vertex indices are out of range.
        """
        if not (0 <= u < self.vertices and 0 <= v < self.vertices):
            raise ValueError("Vertex indices must be within valid range.")
        self.distances[u][v] = weight
        self.next[u][v] = v  # Set the next vertex in the shortest path.

    def run_algorithm(self):
        """
        Run the Floyd-Warshall algorithm to compute shortest paths between all pairs of vertices.
        """
        # Loop over each intermediate vertex.
        for k in range(self.vertices):
            # Loop over each source vertex.
            for i in range(self.vertices):
                # Loop over each destination vertex.
                for j in range(self.vertices):
                    # Check if going through vertex k offers a shorter path.
                    if self.distances[i][k] + self.distances[k][j] < self.distances[i][j]:
                        # Update the shortest path distance.
                        self.distances[i][j] = self.distances[i][k] + self.distances[k][j]
                        # Update the next vertex in the path.
                        self.next[i][j] = self.next[i][k]

    def has_negative_cycle(self):
        """
        Check if the graph contains a negative cycle.

        :return: True if a negative cycle exists, False otherwise.
        """
        for i in range(self.vertices):
            if self.distances[i][i] < 0:
                return True
        return False

    def get_shortest_distance(self, u, v):
        """
        Get the shortest distance between two vertices.

        :param u: The source vertex (0-indexed).
        :param v: The destination vertex (0-indexed).
        :return: The shortest distance, or float('inf') if no path exists.
        :raises ValueError: If vertex indices are out of range.
        """
        if not (0 <= u < self.vertices and 0 <= v < self.vertices):
            raise ValueError("Vertex indices must be within valid range.")
        return self.distances[u][v]

    def reconstruct_path(self, u, v):
        """
        Reconstruct the shortest path between two vertices.

        :param u: The source vertex (0-indexed).
        :param v: The destination vertex (0-indexed).
        :return: A list representing the path from u to v, or an empty list if no path exists.
        :raises ValueError: If vertex indices are out of range.
        """
        if not (0 <= u < self.vertices and 0 <= v < self.vertices):
            raise ValueError("Vertex indices must be within valid range.")

        # If there's no path, return an empty list.
        if self.distances[u][v] == float('inf'):
            return []

        # Initialize the path with the source vertex.
        path = [u]
        while u != v:
            u = self.next[u][v]  # Follow the next pointers to reconstruct the path.
            path.append(u)

        return path


# Example usage:
if __name__ == "__main__":
    # Create a graph with 4 vertices.
    graph = FloydWarshall(4)

    # Add edges with their weights.
    graph.add_edge(0, 1, 5)
    graph.add_edge(0, 3, 10)
    graph.add_edge(1, 2, 3)
    graph.add_edge(2, 3, 1)

    # Run the Floyd-Warshall algorithm.
    graph.run_algorithm()

    # Check for negative cycles.
    if graph.has_negative_cycle():
        print("The graph contains a negative weight cycle.")
    else:
        # Get the shortest distance between vertices 0 and 3.
        print("Shortest distance from 0 to 3:", graph.get_shortest_distance(0, 3))

        # Reconstruct the shortest path from vertex 0 to 3.
        print("Shortest path from 0 to 3:", graph.reconstruct_path(0, 3))

File: course_4/module_1/examples/floyd_warshall.go
package main

import (
	"fmt"
	"math"
)

// FloydWarshall represents a graph and contains methods to compute shortest paths.
type FloydWarshall struct {
	vertices  int
	distances [][]float64
	next      [][]int
}

// NewFloydWarshall creates a new instance of the FloydWarshall algorithm.
func NewFloydWarshall(vertices int) *FloydWarshall {
	distances := make([][]float64, vertices)
	next := make([][]int, vertices)
	for i := 0; i < vertices; i++ {
		distances[i] = make([]float64, vertices)
		next[i] = make([]int, vertices)
		for j := 0; j < vertices; j++ {
			if i == j {
				distances[i][j] = 0
			} else {
				distances[i][j] = math.Inf(1)
			}
			next[i][j] = -1
		}
	}
	return &FloydWarshall{vertices, distances, next}
}

// AddEdge adds an edge to the graph with a specified weight.
func (fw *FloydWarshall) AddEdge(u, v int, weight float64) error {
	if err := fw.validateVertex(u); err != nil {
		return err
	}
	if err := fw.validateVertex(v); err != nil {
		return err
	}
	fw.distances[u][v] = weight
	fw.next[u][v] = v
	return nil
}

// validateVertex checks if a vertex index is valid.
func (fw *FloydWarshall) validateVertex(v int) error {
	if v < 0 || v >= fw.vertices {
		return fmt.Errorf("vertex index %d is out of range 0 to %d", v, fw.vertices-1)
	}
	return nil
}

// RunAlgorithm runs the Floyd-Warshall algorithm to compute shortest paths.
func (fw *FloydWarshall) RunAlgorithm() {
	for k := 0; k < fw.vertices; k++ {
		for i := 0; i < fw.vertices; i++ {
			for j := 0; j < fw.vertices; j++ {
				if fw.distances[i][k]+fw.distances[k][j] < fw.distances[i][j] {
					fw.distances[i][j] = fw.distances[i][k] + fw.distances[k][j]
					fw.next[i][j] = fw.next[i][k]
				}
			}
		}
	}
}

// HasNegativeCycle checks if the graph contains a negative weight cycle.
func (fw *FloydWarshall) HasNegativeCycle() bool {
	for i := 0; i < fw.vertices; i++ {
		if fw.distances[i][i] < 0 {
			return true
		}
	}
	return false
}

// GetShortestDistance returns the shortest distance between two vertices.
func (fw *FloydWarshall) GetShortestDistance(u, v int) (float64, error) {
	if err := fw.validateVertex(u); err != nil {
		return math.Inf(1), err
	}
	if err := fw.validateVertex(v); err != nil {
		return math.Inf(1), err
	}
	return fw.distances[u][v], nil
}

// ReconstructPath reconstructs the shortest path between two vertices.
func (fw *FloydWarshall) ReconstructPath(u, v int) ([]int, error) {
	if err := fw.validateVertex(u); err != nil {
		return nil, err
	}
	if err := fw.validateVertex(v); err != nil {
		return nil, err
	}

	if fw.distances[u][v] == math.Inf(1) {
		return []int{}, nil
	}

	path := []int{u}
	for u != v {
		u = fw.next[u][v]
		if u == -1 {
			return nil, fmt.Errorf("no path exists between the given vertices")
		}
		path = append(path, u)
	}

	return path, nil
}

func main() {
	graph := NewFloydWarshall(4)

	if err := graph.AddEdge(0, 1, 5); err != nil {
		fmt.Println("Error adding edge:", err)
	}
	if err := graph.AddEdge(0, 3, 10); err != nil {
		fmt.Println("Error adding edge:", err)
	}
	if err := graph.AddEdge(1, 2, 3); err != nil {
		fmt.Println("Error adding edge:", err)
	}
	if err := graph.AddEdge(2, 3, 1); err != nil {
		fmt.Println("Error adding edge:", err)
	}

	graph.RunAlgorithm()

	if graph.HasNegativeCycle() {
		fmt.Println("The graph contains a negative weight cycle.")
	} else {
		dist, err := graph.GetShortestDistance(0, 3)
		if err != nil {
			fmt.Println("Error getting shortest distance:", err)
		} else {
			fmt.Printf("Shortest distance from 0 to 3: %v\n", dist)
		}

		path, err := graph.ReconstructPath(0, 3)
		if err != nil {
			fmt.Println("Error reconstructing path:", err)
		} else {
			fmt.Printf("Shortest path from 0 to 3: %v\n", path)
		}
	}
}
